asm/f64:  Updated axpy assembly to wide, pipelined loops.

Kunde21 opened the pull request:
Continuation on gonum/internal#50 after miscommunication.

kortschak committed:
all: remove tests and references to generated code

kortschak committed:
asm/f32: fix build tags

kortschak committed:
asm: fix build retry

Kunde21 committed:
asm/f64: Alignment bug in DivTo fixed with test added.

kortschak committed:
internal/asm/...: remove outdated commments

kortschak committed:
internal/asm/...: add documentation

This documentation is intended to allow gonum developers to see exactly
what an asm function is intended to do. They are not for external code
consumers.

kortschak committed:
internal/asm/...: rename functions

These are unused outside the asm packages and so are safe to rename now.

L1Norm => L1Dist
LinfNorm => LinfDist
AbsSum => L1Norm

kortschak committed:
internal/README: add gddo link and badge

Kunde21 committed:
asm/f64: Update guardIncVector to allow negative increments.

Kunde21 committed:
asm/f64: Added and updated testing helper functions.

Added functios generate test increment sets. Added alignment offset sets for testing. Updated axpy test code to use value sigils in guards.

Kunde21 committed:
asm/f64: Changed axpy benchmarks to better cover mid-sized vectors.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -65,3 +68,48 @@ func nonStridedWrite(x []float64, inc int) bool {
+
+var gs []float64
+
+func randomSlice(n, inc int) []float64 {
```
Make this be `randomSlice(n, inc int, src *rand.Rand) []float64` and provide a `*rand.Rand` to use so tests are deterministic.

Kunde21 replied:
This is only used in Dscal* benchmarks and will be disappearing when I re-work that code, because it's not serving any purpose there that isn't better served by the global benchmark slices. 

I only moved it because I was removing it from the Axpy* benchmarks, for the same reason.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -69,6 +69,10 @@ func nonStridedWrite(x []float64, inc int) bool {
 	return false
 }
 
+var align2 = []struct{ x, y int }{{0, 0}, {0, 1}, {1, 0}, {1, 1}}
```
Can these be in a var block and have a comment explaining what they are for.

kortschak reviewed (changes requested):
Can you squash together some of the related (double tap) commits (for example bccaafc and 4baf769) and also if it's not too much work split the first commit into the changes that are described in the commit message and another for the additional loop reorganisation.

It would be nice to have this rebased onto master rather than master merged, but looking at that and the conflicts that arise, it looks like too much work to bother with.

Kunde21 committed:
asm/f64: Consolidated testing helper functions into asm_test.go

Kunde21 commented:
Could you clarify the loop reorganization you'd like me to pull out (I'm assuming 31ce856 is the commit you're referring)?

kortschak commented:
Sure. The changes in 8bbbf2f46887ef973b17b6a9ec77fbcd5e5b3251 clean up some naming which is really nice, but also split the tail into two branches. What I would have liked here is a completely mechanical change to add the register aliasing, then manual changes cleaning up comments and then manual flow/structure changes. If I know changes are mechanical (plus trivial additions like the `#define`s I can review that easily, and spend relatively little time on the review of any actual changes in their focused commits.

The main reason I'm asking for this is that reviewing asm takes me probably about 5-10x as long as reviewing Go/Perl/Python/C, so reducing the amount of non-trivial code change diffs really helps.

Kunde21 commented:
Gotcha.  I'll try to pull that apart tonight.

Kunde21 committed:
asm/f64: Defined macros for named variable registers in axpy* code.

Kunde21 committed:
asm/f64:  Cleaned up axpy* assembly code.

Used 3xIncrement registers to consolidate pointer arithmetic to the
end of the main loop.  Moved pipelining register calculations to after
short-vector test. Added two-wide tail code block.

Kunde21 commented:
Let me know if that helps.  I pulled the mechanical changes into 724bdb1

kortschak commented:
Thanks. That's better. I should be able to take a serious look at this over the weekend.

Kunde21 commented:
Do you want me to similarly scrub up the scal* asm/test code before then?

kortschak commented:
That would be great. Thanks.

kortschak commented on asm/f64/axpyunitaryto_amd64.s:
```diff
@@ -51,34 +51,36 @@
+	XORQ   IDX, IDX            // IDX = 0
+	MOVSD  alpha+24(FP), ALPHA
+	SHUFPD $0, ALPHA, ALPHA    // ALPHA := { alpha, alpha }
+	MOVQ   Y_PTR, TAIL         // Check memory alignment
```
These should happen in `thiscommit~1`.

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -0,0 +1,162 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
```
This is all new code, so 2017.

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -0,0 +1,162 @@
+
+var (
+	a = float64(2)
+	x = make([]float64, 1000000)
```
Make a const for the size here.

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -0,0 +1,162 @@
+		{"NaiveAxpyUnitary", naiveaxpyu},
+	}
+	for _, tst := range tests {
+		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 1e4, 5e4, 1e5, 5e5} {
```
Maybe just ~0.5 log10 spacing is enough: 1, 3, 10, 30, 1e2, 3e2, 1e3, 3e3...

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -0,0 +1,162 @@
+		{"NaiveAxpyUnitaryTo", naiveaxpyut},
+	}
+	for _, tst := range tests {
+		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 1e4, 5e4, 1e5, 5e5} {
```
Same here.

kortschak commented:
This breaks blas, but reviewing anyway.

```
$ go test github.com/gonum/blas/native
unexpected fault address 0xc420100000
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc420100000 pc=0x4f501b]

goroutine 36 [running]:
runtime.throw(0x5b7fd2, 0x5)
	/home/daniel/go/src/runtime/panic.go:596 +0x95 fp=0xc4200b85a8 sp=0xc4200b8588
runtime.sigpanic()
	/home/daniel/go/src/runtime/signal_unix.go:297 +0x28c fp=0xc4200b85f8 sp=0xc4200b85a8
github.com/gonum/internal/asm/f64.AxpyUnitaryTo(0xc4200fbab8, 0x0, 0x9, 0x3ff5f15f15f15f17, 0xc4200fda40, 0x0, 0x0, 0xc4200fbab8, 0x0, 0x9, ...)
	/home/daniel/src/github.com/gonum/internal/asm/f64/axpyunitaryto_amd64.s:90 +0x7b fp=0xc4200b8600 sp=0xc4200b85f8
github.com/gonum/blas/native.Implementation.Dtrsm(0x8e, 0x79, 0x6f, 0x83, 0x4, 0x3, 0x4008000000000000, 0xc4200fda40, 0x9, 0x9, ...)
	/home/daniel/src/github.com/gonum/blas/native/level3double.go:185 +0x1058 fp=0xc4200b8820 sp=0xc4200b8600
github.com/gonum/blas/native.(*Implementation).Dtrsm(0x6bd068, 0x8e, 0x79, 0x6f, 0x83, 0x4, 0x3, 0x4008000000000000, 0xc4200fda40, 0x9, ...)
	<autogenerated>:76 +0x117 fp=0xc4200b88a8 sp=0xc4200b8820
github.com/gonum/blas/testblas.DtrsmTest(0xc4200f60d0, 0x67f660, 0x6bd068)
	/home/daniel/src/github.com/gonum/blas/testblas/dtrsm.go:806 +0x7ecb fp=0xc4200b9f78 sp=0xc4200b88a8
github.com/gonum/blas/native.TestDtrsm(0xc4200f60d0)
	/home/daniel/src/github.com/gonum/blas/native/level3double_test.go:22 +0x5b fp=0xc4200b9fa8 sp=0xc4200b9f78
testing.tRunner(0xc4200f60d0, 0x5c3b68)
	/home/daniel/go/src/testing/testing.go:657 +0x96 fp=0xc4200b9fd0 sp=0xc4200b9fa8
runtime.goexit()
	/home/daniel/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4200b9fd8 sp=0xc4200b9fd0
created by testing.(*T).Run
	/home/daniel/go/src/testing/testing.go:697 +0x2ca

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42005e680, 0x5b8b1d, 0x9, 0x5c3b68, 0xc420043d01)
	/home/daniel/go/src/testing/testing.go:698 +0x2f4
testing.runTests.func1(0xc42005e680)
	/home/daniel/go/src/testing/testing.go:882 +0x67
testing.tRunner(0xc42005e680, 0xc420043de0)
	/home/daniel/go/src/testing/testing.go:657 +0x96
testing.runTests(0xc42000adc0, 0x69c100, 0x24, 0x24, 0x7eff8bb22000)
	/home/daniel/go/src/testing/testing.go:888 +0x2c1
testing.(*M).Run(0xc420043f20, 0xc420043f20)
	/home/daniel/go/src/testing/testing.go:822 +0xfc
main.main()
	github.com/gonum/blas/native/_test/_testmain.go:514 +0xf7
FAIL	github.com/gonum/blas/native	0.005s
```

kortschak commented on asm/f64/axpyinc_amd64.s:
```diff
@@ -131,8 +131,8 @@ tail_two: // do {
 
-	CMPQ TAIL, $0
-	JE   end      // if TAIL == 0 { goto end }
+	ANDQ $1, TAIL
```
Can you explain the bug and fix in the commit message please. This seems odd; is TAIL never even?

Kunde21 replied:
This commit got squashed, but I'll explain it here. 

The two-wide block tests the 2's bit using `MOVQ TAIL, LEN` and `SHRQ $1, LEN`.  But `TAIL` isn't modified, so we use `ANDQ $1, TAIL` to test only the 1's bit.  `ANDQ` sets the `ZF` flag if the result is zero, so this is (with the following conditional jump) an `if TAIL & 1 ==0 { return }` block.  The `CMPQ TAIL, $0` version gives false positives for any tail values, even or odd, which clobbers the trailing guard in tests.

kortschak replied:
Ah, I missed the shift on LEN. Thanks.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -65,3 +68,48 @@ func nonStridedWrite(x []float64, inc int) bool {
+	x, y int
+}
+
+func genInc(inc []int) []incSet {
```
Please add a doc comment to these two helpers and their associated types. I would also try to think of better names so that the tests are more fluently read. I'd recommend `newIncSet` and `newIncToSet` for the functions.

For both, you may want to use the variadic form for `inc`.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -65,3 +68,48 @@ func nonStridedWrite(x []float64, inc int) bool {
+	return is
+}
+
+var gs []float64
```
What is this for?

Kunde21 replied:
I'm not sure.  Seemed like an artifact to me.  

https://github.com/gonum/internal/blob/master/asm/f64/scal_test.go#L213

Kunde21 committed:
asm/f64: Zero-length exit for axpyunitaryto with regression test.

Kunde21 commented:
Not sure how, but the zero-length fix commit got lost in the rebase.  That's my fault.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -65,3 +68,48 @@ func nonStridedWrite(x []float64, inc int) bool {
+	dst, x, y int
+}
+
+func genIncTo(inc []int) []incToSet {
```
This is not right. With suggestions from above:

```
func newIncToSet(inc ...int) []incToSet {
	n := len(inc)
	is := make([]incToSet, n*n*n)
	for i, dst := range inc {
		for x := range inc {
			for y := range inc {
				is[i*n*n+x*n+y] = incToSet{dst, inc[x], inc[y]}
			}
		}
	}
	return is
}
```

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -69,6 +69,12 @@ func nonStridedWrite(x []float64, inc int) bool {
 }
 
+var ( // Offset sets for testing alignment handling in Unitary assembly functions.
+	align2 = []struct{ x, y int }{{0, 0}, {0, 1}, {1, 0}, {1, 1}}
```
```
align2 = newIncSet(0, 1)
align3 = newIncToSe(0, 1)
```

Also, squash `thiscommit` with `thiscommit~1` to make reason for existence clear.

kortschak commented:
Thanks.

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
-		if !equalStrided(test.want, y, 1) {
-			t.Errorf("%v: unexpected result:\nwant: %v\ngot: %v", prefix, test.want, y)
+func TestAxpyUnitary(t *testing.T) {
+	var xGdVal, yGdVal float64 = -1, 0.5
```
These can be `const`.

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
-		if !equalStrided(test.want, x, 1) {
-			t.Errorf("%v: unexpected result:\nwant: %v\ngot: %v", prefix, test.want, x)
+func TestAxpyUnitaryTo(t *testing.T) {
+	var dstGdVal, xGdVal, yGdVal float64 = 1, -1, 0.5
```
`const`

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
-			t.Errorf("%v: unexpected result:\nwant: %v\ngot: %v", prefix, test.want, y)
+func TestAxpyUnitary(t *testing.T) {
+	var xGdVal, yGdVal float64 = -1, 0.5
+	for i, tt := range axpyTests {
```
s/tt/test/g; in tests `test.foo` is a convention here for referring to test records.

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
-				y, yFront, yBack := newGuardedVector(test.yData, incY)
-				AxpyInc(test.alpha, x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
+func TestAxpyInc(t *testing.T) {
+	var xGdVal, yGdVal float64 = -1, 0.5
```
`const`

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
+	gdLn := 4
+	for i, test := range axpyTests {
+		n := len(test.x)
+		for _, inc := range genInc([]int{-7, -4, -3, -2, -1, 1, 2, 3, 4, 7}) {
```
`newIncSet(-7, -4, -3, -2, -1, 1, 2, 3, 4, 7)`

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
+	gdLn := 4
+	for i, tt := range axpyTests {
+		n := len(tt.x)
+		for _, inc := range genIncTo([]int{-7, -4, -3, -2, -1, 1, 2, 3, 4, 7}) {
```
`newIncToSet(-7, -4, -3, -2, -1, 1, 2, 3, 4, 7)`

kortschak commented on asm/f64/axpy_test.go:
```diff
@@ -6,533 +6,262 @@ package f64
+	msgGuard = "%v: Guard violated in %s vector %v %v"
+)
 
+var axpyTests = []struct {
```
Squash `thiscommit` with `thiscommit~1` and `thiscommit~2`.

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -39,7 +39,7 @@ func BenchmarkAxpyUnitary(t *testing.B) {
 	}
 	for _, tst := range tests {
-		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 1e4, 5e4, 1e5, 5e5} {
+		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 500, 1e3, 5e3, 1e4, 5e4} {
```
Use ~0.5 log10 steps.

kortschak commented on asm/f64/benchAxpy_test.go:
```diff
@@ -66,7 +66,7 @@ func BenchmarkAxpyUnitaryTo(t *testing.B) {
 	}
 	for _, tst := range tests {
-		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 1e4, 5e4, 1e5, 5e5} {
+		for _, ln := range []uintptr{1, 2, 3, 4, 5, 10, 100, 500, 1e3, 5e3, 1e4, 5e4} {
```
Same here.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -79,6 +149,7 @@ type incSet struct {
 	x, y int
 }
 
+// genInc will generate all (x,y) combinations of the input increment set.
```
Thanks.

kortschak commented on asm/f64/asm_test.go:
```diff
@@ -94,6 +165,7 @@ type incToSet struct {
 	dst, x, y int
 }
 
+// genIncTo will generate all (dst,x,y) combinations of the input increment set.
```
Thanks.

Kunde21 commented:
DAMN IT ALL TO HELL!!!  

The instructions for fixing conflicts don't warn that it also merges the PR.

kortschak commented:
Please revert that.

I'd recommend you avoid using anything other than the cl git commands locally.

Kunde21 commented:
It wouldn't allow me to fix the conflicts on here.  

At this point, we should rebase my whole debacle out.

kortschak commented:
That's not possible. We don't allow force changes to master.

kortschak commented:
Reverted by abbe1115275b8ef5f1c4b73e2360c9e8177edcb2

Kunde21 commented:
It's not giving me the option to create a revert.

kortschak commented:
Yeah, I noticed that, so I did it with git.

Kunde21 commented:
Thank you.  

I'm really trying not to break everything (he says, strewn remains of everything in the background).

kortschak commented:
No worries. No one died.
